File: modules/agent/src/services/technology_service.py
"""Service layer for technology business logic."""

from typing import Optional, List

from fastapi import HTTPException

from repositories.technology_repository import (
    find_all_technologies,
    find_technology_by_id,
    create_technology as create_technology_repo,
)


async def get_all_technologies(category: Optional[str] = None) -> List:
    """Get all technologies, optionally filtered by category."""
    return await find_all_technologies(category=category)


async def get_technology(tech_id: int):
    """Get a single technology by ID."""
    tech = await find_technology_by_id(tech_id)
    if not tech:
        raise HTTPException(status_code=404, detail="Technology not found")
    return tech


async def create_technology(name: str, category: str, vendor: Optional[str] = None):
    """Create a new technology."""
    return await create_technology_repo(name=name, category=category, vendor=vendor)

File: modules/agent/internal/services/technology_service.go
package services

import (
	"errors"

	"agent/internal/repositories"
	"agent/internal/schemas"
	"agent/internal/serializers"
)

var ErrTechnologyNotFound = errors.New("technology not found")

// TechnologyService handles technology business logic
type TechnologyService struct {
	techRepo *repositories.TechnologyRepository
}

// NewTechnologyService creates a new technology service
func NewTechnologyService(techRepo *repositories.TechnologyRepository) *TechnologyService {
	return &TechnologyService{techRepo: techRepo}
}

// GetAllTechnologies returns all technologies, optionally filtered by category
func (s *TechnologyService) GetAllTechnologies(category *string) ([]serializers.TechnologyResponse, error) {
	techs, err := s.techRepo.FindAll(category)
	if err != nil {
		return nil, err
	}
	return serializers.TechnologiesToResponse(techs), nil
}

// GetTechnology returns a technology by ID
func (s *TechnologyService) GetTechnology(id int64) (*serializers.TechnologyResponse, error) {
	tech, err := s.techRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if tech == nil {
		return nil, ErrTechnologyNotFound
	}
	resp := serializers.TechnologyToResponse(tech)
	return &resp, nil
}

// CreateTechnology creates a new technology
func (s *TechnologyService) CreateTechnology(input schemas.TechnologyCreate) (*serializers.TechnologyResponse, error) {
	repoInput := repositories.TechnologyCreateInput{
		Name:     input.Name,
		Category: input.Category,
		Vendor:   input.Vendor,
	}
	tech, err := s.techRepo.Create(repoInput)
	if err != nil {
		return nil, err
	}
	resp := serializers.TechnologyToResponse(tech)
	return &resp, nil
}
